### ISSUE TITLE
New rule: Warn/Error on two or more consecutive uppercase characters in a word

### ISSUE THREAD

**Issue by carlosduclos** [eslint/eslint#9538]
Please describe what the rule should do:
Warn or error if a word contains two or more consecutive uppercase characters.

What category of rule is this? (place an "X" next to just one item)

[ ] Enforces code style [X] Warns about a potential error [ ] Suggests an alternate way of doing something [ ] Other (please specify:)

Provide 2-3 code examples that this rule will warn about:
* bad:
hostADdress -> The developer most likely meant "hostAddress"

* bad:
hostQUalifiedName: The developer most likely meant hostQualifiedName

This code can be rearranged and written in other ways, this is just a quick example.
```javascript
function checkConsecutiveUpperLetters(variables) {
    let faulty = [];
    variables.forEach(function(variable) {
        if (variable.name.match(/[ABCDEFGHIJKLMNOPQRSTUVWXYZ]{2,}/)) {
            faulty.push(variable.name);
        }
    });
    return faulty;
};

module.exports = {
    meta: {
        docs: {
            description: 'No consecutive uppercase letters',
            category: 'Possible errors',
            recommended: true
        },
        schema: [] // no options
    },
    create: function(context) { 
        return {
            onCodePathStart: function(codePath, node) {
                const consecutiveUpperLettersWords = checkConsecutiveUpperLetters(context.getDeclaredVariables(node));
                if (consecutiveUpperLettersWords.length > 0) {
                    consecutiveUpperLettersWords.forEach(function(word) {
                        // report
                    });    
                }
            }
        };
    }
};
```

Why should this rule be included in ESLint (instead of a plugin)?
This rule will be helpful to many people. Misspelling is one of the most common causes of bugs.

I'm working on a PR if this is accepted.

**Comment by Aladdin-ADD** [comment]
thanks! we have a rule `id-match` can work for this. so I'm not sure we should add a new rule.

**Comment by carlosduclos** [comment]
I'll be glad to use an existing rule if it already exists. I'll look into the id-match rule, and I'll continue working on the pull request just in case.

**Comment by carlosduclos** [comment]
It seems that this particular type of rule is one of the cases where this plugin should not be used (as mentioned in the docs).
For one, it is quite complicated to build a regexp that matches this kind of negative matching, and even if you cover all the cases, the performance is really horrible.

**Comment by not-an-aardvark** [comment]
Could the match regex just be `/[A-Z]{2,}/`? It doesn't seem to me that the performance would be very bad.

In my opinion, the rule that you're proposing is a fairly user-specific preference, so I'm not sure I would support adding a rule for it to ESLint core. The goal of the `id-match` rule (and some others like it) is to be sufficiently configurable that users can enforce checks on identifier names without needing specific rules to be added for each case.

**Comment by platinumazure** [comment]
@not-an-aardvark I think the problem is that `id-match` regexes are for allowed identifiers, not blocked identifiers. So really, the user would have to negate that regex, which is non-trivial.

I think we could reasonably support an enhancement to the rule where instead of just providing a regex for valid identifier patterns, we could also accept one or more regexes representing disallowed identifier patterns.

**Comment by not-an-aardvark** [comment]
Oh, I see. Since any regular expression can be negated with another regular expression, an option like that seems a bit redundant to me. But I'm not strongly opposed to it if others think that would make things easier.

I do think it would be better for users to use a general solution rather than creating a core rule specifically for double-capital-letters.

**Comment by carlosduclos** [comment]
As @platinumazure mentions, the major problem with using id-match is the fact that is designed to match and not to reject. It might seem a minor thing, but in the case at hand it requires a very complex regexp to get it right for most cases (and it seems impossible to get it right for all of them):
```javascript
{
  "rules": {
     "id-match": ["error", "^_?[A-Z]?[a-z0-9_.]+(?![A-Z]{2,})"]
   }
}
```
Instead by using rejection logic, the regexp would be as simple as:
```javascript
"[A-Z]{2,}"
```
I see the point in generalizing this and I propose the following. The rejection logic is the same, whether it is to find two consecutive uppercase characters or any other pattern, so I propose to write a rule called "id-reject" which will reject identifiers based on the given patterns.

I have thought about reusing id-match and adding a flag to change the behavior, but it is completely valid for somebody to want to both match and reject at the same time.

I'm going to generalize my work and send a PR when it is ready to be shown.

**Comment by not-an-aardvark** [comment]
Another option would be to use [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax):

```json
{
  "rules": {
    "no-restricted-syntax": ["error", "Identifier[name=/[A-Z]{2,}/]"]
  }
}
```

> I'm going to generalize my work and send a PR when it is ready to be shown.

Thanks for contributing! You're certainly free to create a PR, although it can sometimes be better to wait until the details are fully worked out in the corresponding issue, so that you don't end up having to redo/scrap your work if the proposal evolves after discussion.

---

Personally, I feel like we already have several ways to accomplish the goal of disallowing specified identifier patterns (either by constructing an inverted regex in `id-match`, or by using a selector in `no-restricted-syntax`), so I'm not convinced that adding another way is worthwhile.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
